---
title: "HpyA_arCOG"
author: "Saaz Sakrikar"
date: "7/16/2021"
output: html_document
---

Finds enriched COGs for HpyA
Uses COG file  Hbt_arCOGs.csv
  
Input  has  a list of Oldgenename.
Input is generated by DeSeq2 and then modified manually. The gene names are under the column "Oldgenename".

1st output is a list  of categories with pval<0.05 (this number can be adjusted)
2nd output is a list of genes within each enriched category
3rd output is the input table of differentially expressed genes (DEG), with COG labels added to each gene if available.
4th and 5th outputs are COG enrichments based on category - reduced salt only, and DEGs in both reduced and optimal salt.

All 3 outputs are visible here in this R notebook and also outputted as excel files:
"allCOGpvals.xlsx", "COGenrichedgenes.xlsx", "RNASeqtablewithCOGs.xlsx"

# Load required packages
```{r libraries}
library(readxl)
library(bazar)
library(openxlsx)
library(ggplot2)
library(tidyr)
```

```{r set-options, echo=FALSE, cache=FALSE}
options(width = 300)
```

# Load arCOG table (from Makarova et al., 2015 DOI: 10.3390/life5010818)
```{r inputs}

cogfile=read.csv("Hbt_arCOGs.csv")
RNASeqoutput=read_excel("hpyA_RNASeqhits_improved.xlsx")
#RNASeqoutput=read_excel("June2021_TableS4_diffexpr-AKS-nonredundant-noCombinatorial.xlsx", sheet ="KO_sigresults")

namelist=RNASeqoutput$Oldgenename
namelist=as.character(namelist)

l=length(namelist)

pvalue=0.05
Name="Hsal_hpyA"
```


#Find enriched categories across all DEGs
```{r subsetting, warning=FALSE}
cogfile$funclass_name=as.factor(cogfile$funclass_name)
cogs= subset(cogfile, is.element(cogfile$GeneName, namelist)==TRUE)
clust= summary(cogs$funclass_name)

resm <- matrix(0,length(clust),3)


res <-data.frame(resm)
rownames(res) <- names(clust)
colnames(res) <- c("probability", "expect","count")

all=summary(cogfile$funclass_name)
totaln=sum(all)

for (i in 1:length(clust)){
  res[i,1] <- phyper(clust[i], all[i], totaln-all[i],length(namelist) ,lower.tail=F)
  res[i,2] <- all[i]*(length(namelist)/totaln)
  res[i,3] <- clust[i]
}

res$padj <- p.adjust(res$probability, method = "BH")
res.signif  <- subset(res,probability<=pvalue&count!=0)
res.signif$padj <- p.adjust(res.signif$probability, method  = "BH")
```

# Find all genes within these enriched categories.
```{r listsofenrichedgenes}
funclasses=rownames(res)
l2=length(funclasses)
tmpnumber=dim(cogs)[2]
Listofgenes=as.data.frame(matrix(nrow=1,ncol=tmpnumber))
colnames(Listofgenes)=colnames(cogs)


for (i in (1:l2)) {
  relclass=funclasses[i]
  relpval=res$padj[i]
  if (relpval<pvalue) {
    addedgenes=subset(cogs,is.element(cogs$funclass_name,relclass))
    Listofgenes=rbind(Listofgenes,addedgenes)
  }
}

Listofgenes=Listofgenes[-1,]
#Listofgenes=Listofgenes[,c(27,3,4,10,19,20,21)]

head(Listofgenes)
```
# Adding COG information to list of DEGs to make supplementary table S4
```{r addCOGstogenelist,warning=FALSE}

InputwithCOGs=RNASeqoutput
classnamevector=as.character(vector(length=l))
classsymbolvector=as.character(vector(length=l))
annotationvector=as.character(vector(length=l))

for (i in (1:l)) {
  relgene=namelist[i]
  relindex=grep(relgene,cogfile$GeneName,ignore.case=TRUE)
  if (!is.empty(relindex)) {
    classnamevector[i]=as.character(cogfile$funclass_name[relindex])
    classsymbolvector[i]=as.character(cogfile$functional_class[relindex])
    annotationvector[i]=as.character(cogfile$arcog_name[relindex])
  }
}

InputwithCOGs=cbind(InputwithCOGs,classsymbolvector,classnamevector,annotationvector)

#InputwithCOGs=InputwithCOGs[,c(2,4,8,15)]
colnames(InputwithCOGs)[15:17]=c("functional_class","funclass_name","arcog_name")

head(InputwithCOGs)
```

#Enrichments among genes diff expressed in reduced salt only
```{r lowsaltgenes,warning=FALSE}
lowsaltgenes=RNASeqoutput[RNASeqoutput$Optimal_padj==1,]
lowsaltgenes=lowsaltgenes$Oldgenename

cogfile$funclass_name=as.factor(cogfile$funclass_name)
cogslow= subset(cogfile, is.element(cogfile$GeneName, lowsaltgenes)==TRUE)
l_low=dim(cogslow)[1]
clustlow= summary(cogslow$funclass_name)

resmlow<- matrix(0,length(clustlow),3)
reslow <-data.frame(resmlow)
rownames(reslow) <- names(clustlow)
colnames(reslow) <- c("probability", "expect","count")
all=summary(cogfile$funclass_name)
totaln=sum(all)

for (i in 1:length(clustlow)){
  reslow[i,1] <- phyper(clustlow[i], all[i], totaln-all[i], l_low,lower.tail=F)
  reslow[i,2] <- all[i]*(l_low/totaln)
  reslow[i,3] <- clustlow[i]
}

reslow$padj <- p.adjust(reslow$probability, method = "BH")
reslow.signif  <- subset(reslow,probability<=pvalue&count!=0)
reslow.signif$padj <- p.adjust(reslow.signif$probability, method = "BH")
```

#get list of genes in  enriched categories for reduced salt conditions

```{r listsofenrichedgenes}
funclasses=rownames(reslow.signif)
l2=length(funclasses)
tmpnumber=dim(cogs)[2]
Listoflowgenes=as.data.frame(matrix(nrow=1,ncol=tmpnumber))
colnames(Listoflowgenes)=colnames(cogs)


for (i in (1:l2)) {
  relclass=funclasses[i]
  relpval=reslow.signif$padj[i]
  if (relpval<pvalue) {
    addedgenes=subset(cogs,is.element(cogs$funclass_name,relclass))
    Listoflowgenes=rbind(Listoflowgenes,addedgenes)
  }
}

Listoflowgenes=Listoflowgenes[-1,]
#Listofgenes=Listofgenes[,c(27,3,4,10,19,20,21)]

head(Listoflowgenes)
```



#Enrichments among genes diff expr in both optimal and reduced salt
```{r reduced_optimal_genes,warning=FALSE}
redoptgenes=RNASeqoutput[RNASeqoutput$Optimal_padj<1,]
redoptgenes=redoptgenes[redoptgenes$Low_padj<1,]
redoptgenes=redoptgenes$Oldgenename

cogfile$funclass_name=as.factor(cogfile$funclass_name)
cogsredopt= subset(cogfile, is.element(cogfile$GeneName, redoptgenes)==TRUE)
l_lredopt=dim(cogsredopt)[1]
clustredopt= summary(cogsredopt$funclass_name)

resmredopt<- matrix(0,length(clustredopt),3)
resredopt <-data.frame(resmredopt)
rownames(resredopt) <- names(clustredopt)
colnames(resredopt) <- c("probability", "expect","count")
all=summary(cogfile$funclass_name)
totaln=sum(all)

for (i in 1:length(clustredopt)){
  resredopt[i,1] <- phyper(clustredopt[i], all[i], totaln-all[i], l_lredopt,lower.tail=F)
  resredopt[i,2] <- all[i]*(l_lredopt/totaln)
  resredopt[i,3] <- clustredopt[i]
}
resredopt$padj <- p.adjust(resredopt$probability, method = "BH")
resredopt.signif <- subset(resredopt,probability<=pvalue&count!=0)
resredopt.signif$padj <- p.adjust (resredopt.signif$probability, method = "BH")
```

```{r listsofenrichedgenes}
funclasses=rownames(resredopt.signif)
l2=length(funclasses)
tmpnumber=dim(cogs)[2]
Listofsharedgenes=as.data.frame(matrix(nrow=1,ncol=tmpnumber))
colnames(Listofsharedgenes)=colnames(cogs)


for (i in (1:l2)) {
  relclass=funclasses[i]
  relpval=resredopt.signif$padj[i]
  if (relpval<pvalue) {
    addedgenes=subset(cogs,is.element(cogs$funclass_name,relclass))
    Listofsharedgenes=rbind(Listofsharedgenes,addedgenes)
  }
}

Listofsharedgenes=Listofsharedgenes[-1,]
#Listofgenes=Listofgenes[,c(27,3,4,10,19,20,21)]

head(Listofsharedgenes)
```

```{r output}
write.xlsx(InputwithCOGs,paste(Name,"RNASeqtablewithCOGs.xlsx",sep="_"),row.names=FALSE, overwrite = TRUE)
write.xlsx(Listofgenes,paste(Name,"COGenrichedgenes-padj.xlsx",sep="_"),row.names=FALSE, overwrite = TRUE)
write.xlsx(Listoflowgenes,paste(Name,"COGenrichedgenes_lowsalt-padj.xlsx",sep="_"),row.names=FALSE, overwrite = TRUE)
write.xlsx(Listofsharedgenes,paste(Name,"COGenrichedgenes_optlow-padj.xlsx",sep="_"),row.names=FALSE, overwrite = TRUE)
write.xlsx(res.signif,paste(Name,"allCOGpvals-padj.xlsx",sep="_"),row.names=TRUE, overwrite = TRUE)
write.xlsx(reslow.signif,paste(Name,"reducedonly_COGpvals-padj.xlsx",sep="_"),row.names=TRUE,overwrite = TRUE)
write.xlsx(resredopt.signif,paste(Name,"reducedandoptimal_COGpvals-padj.xlsx",sep="_"),row.names=TRUE, overwrite = TRUE)
```
#Make a graph of functional enrichments
```{r}
#import tab "RNAseq_Enrichedcategories" from Supplementary Table S5
#cog.enr <- read.delim("arCOGs-forR.txt", sep = "\t")
cog.enr <- read.delim("arCOGs-forR-padj.txt", sep = "\t")

pdf("arCOG.bargraph-padj.pdf")
ggplot(cog.enr, aes(x = reorder (COGdesignation, count), y = count, fill = neg.logpadj)) +
  geom_bar(stat = "identity") + 
  coord_flip() +
  theme_classic(base_size = 30) +
  scale_fill_gradient(low = "grey", high = "black", limits = c(0,4)) +
  labs(colour = "neg.logpadj", x = "arCOG category", y = "number of genes")


dev.off()
  
```