---
title: "RPKM_visualisation3_Aug17"
author: "Saaz Sakrikar"
date: "7/31/2021"
output: html_document
---


Visualisation of ChIP-Seq peaks near differentially expressed genes.

Input is perbase depth data generated by bedtools, for RNA-Seq and ChIP-Seq.

```{r setup,message=FALSE}
library(readxl)
library(bazar)
library(ggplot2)
library(zoo)

Gff=read_excel("Hbt_gff.xlsx",col_names=FALSE)

#WT in reduced salt RNA-Seq per-base counts
WTreps=4
WTPerbase1=read.table("SS2_perbase.txt")
WTPerbase2=read.table("SS4_perbase.txt")
WTPerbase3=read.table("SSB_perbase.txt")
WTPerbase4=read.table("SSF_perbase.txt")

#KO in reduced salt RNA-Seq per-base counts
KOreps=4
KOPerbase1=read.table("SS6_perbase.txt")
KOPerbase2=read.table("SS8_perbase.txt")
KOPerbase3=read.table("SSH_perbase.txt")
KOPerbase4=read.table("SSJ_perbase.txt")

#Location of 16S and 23S rRNA genes, to be excluded.
Region16S=c(1875505:1876977)
Region23S=c(1877501:1880414)

#HpyA-HA in reduced salt ChIP-Seq IP and input
Reduced_IP=read.table("hpyA_M3_IP_log_perbase.txt")
Reduced_WCE=read.table("hpyA_M3_WCE_log_perbase.txt")

#HpyA-HA in optimal salt ChIP-Seq IP and input
Optimal_IP=read.table("hpyA_CM3_IP_log_perbase.txt")
Optimal_WCE=read.table("hpyA_CM3_WCE_log_perbase.txt")

#Genes of interest and replicon where they are found
genelist=c("VNG_RS11280","VNG_RS11285","VNG_RS11290","VNG_RS11295","VNG_RS11300","VNG_RS11305")
chrname="NC_002608.1"

#Visualisation info: "buffer" around genes of interest to see data on either side
buffer=0
```

Combine different replicates into one matrix, 
```{r preliminaries}

l=length(genelist)

#Combining replicates  
WT_exprcombined=cbind(WTPerbase1,WTPerbase2,WTPerbase3,WTPerbase4)
WT_exprcombined=WT_exprcombined[,c(1,2,3,6,9,12)]
colnames(WT_exprcombined)=c("Chr","Loc","Rep1","Rep2","Rep3","Rep4")
rm(WTPerbase1,WTPerbase2,WTPerbase3,WTPerbase4)

KO_exprcombined=cbind(KOPerbase1,KOPerbase2,KOPerbase3,KOPerbase4)
KO_exprcombined=KO_exprcombined[,c(1,2,3,6,9,12)]
colnames(KO_exprcombined)=c("Chr","Loc","Rep1","Rep2","Rep3","Rep4")
rm(KOPerbase1,KOPerbase2,KOPerbase3,KOPerbase4)


#Setup ChIP-Seq data
Reduced_IP=Reduced_IP[Reduced_IP$V1==paste(chrname),]
Reduced_WCE=Reduced_WCE[Reduced_WCE$V1==paste(chrname),]

Optimal_IP=Optimal_IP[Optimal_IP$V1==paste(chrname),]
Optimal_WCE=Optimal_WCE[Optimal_WCE$V1==paste(chrname),]

#Getting gene location info for genes of interest.
Gff_genesonly=Gff[Gff$...3=="gene",]
Gff_genesonly=Gff_genesonly[Gff_genesonly$...1==paste(chrname),]

genelocs=Gff_genesonly[,4:5]
genelocs=as.data.frame(genelocs)
genelocs[,1]=as.numeric(genelocs[,1])
genelocs[,2]=as.numeric(genelocs[,2])
genelocsmeans=rowMeans(genelocs)
genelengths=genelocs[,2]-genelocs[,1]

outputmat=as.data.frame(matrix(data=NA,nrow=l,ncol=6))
colnames(outputmat)=c("Genelocation","Genestart", "Genestop","Name")

for (i in (1:l)) {
  geneindex=grep(genelist[i],Gff_genesonly$...9)
  outputmat[i,1]=genelocsmeans[geneindex]
  outputmat[i,2]=genelocs$...4[geneindex]
  outputmat[i,3]=genelocs$...5[geneindex]
  outputmat[i,4]=genelist[i]
}

#Find entire region of interest - from start of 1st gene to end of last gene.
Regionstart=min(outputmat$Genestart)
Regionstop=max(outputmat$Genestop)
```

Refine RNA-Seq data: convert raw per-base counts to relative counts by dividing by total  (non-rRNA) counts. Select the region of interest for visualisation.

```{r assignlocstocounts}

#Get total and rRNA counts, to obtain a number for normalisation of individual loci, and exclude rRNA from the total.
WTreadtotals=as.data.frame(matrix(nrow=WTreps,ncol=5))
colnames(WTreadtotals)=c("Totalcounts","Counts16S","Counts23S","rRNAtotal","nonrRNAcounts")
KOreadtotals=as.data.frame(matrix(nrow=KOreps,ncol=5))
colnames(KOreadtotals)=c("Totalcounts","Counts16S","Counts23S","rRNAtotal","nonrRNAcounts")

for (i in (1:WTreps)) {
  WTreadtotals$Totalcounts[i]=sum(WT_exprcombined[,(i+2)])
  WTreadtotals$Counts16S[i]=sum(WT_exprcombined[Region16S,(i+2)])
  WTreadtotals$Counts23S[i]=sum(WT_exprcombined[Region23S,(i+2)])
}

for (i in (1:KOreps)) {
  KOreadtotals$Totalcounts[i]=sum(KO_exprcombined[,(i+2)])
  KOreadtotals$Counts16S[i]=sum(KO_exprcombined[Region16S,(i+2)])
  KOreadtotals$Counts23S[i]=sum(KO_exprcombined[Region23S,(i+2)])
}

WTreadtotals$rRNAtotal=WTreadtotals$Counts16S+WTreadtotals$Counts23S
WTreadtotals$nonrRNAcounts=WTreadtotals$Totalcounts-WTreadtotals$rRNAtotal

KOreadtotals$rRNAtotal=KOreadtotals$Counts16S+KOreadtotals$Counts23S
KOreadtotals$nonrRNAcounts=KOreadtotals$Totalcounts-KOreadtotals$rRNAtotal

#Choose data from relevant replicon (chromosome  or plasmid)
WTRegionreads=WT_exprcombined[WT_exprcombined$Chr==paste(chrname),]
KORegionreads=KO_exprcombined[KO_exprcombined$Chr==paste(chrname),]

#Select data from the genes of interest and surrounding region.
startindex=which(WTRegionreads[,2]==Regionstart)
startindex=startindex-buffer
stopindex=which(WTRegionreads[,2]==Regionstop)
stopindex=stopindex+buffer

WTRegionreads=WTRegionreads[startindex:stopindex,]
KORegionreads=KORegionreads[startindex:stopindex,]

#Normalisation by total counts
for (i in (1:WTreps)) {
  WTRegionreads[,(i+2)]=WTRegionreads[,(i+2)]/WTreadtotals$nonrRNAcounts[i]
  KORegionreads[,(i+2)]=KORegionreads[,(i+2)]/KOreadtotals$nonrRNAcounts[i]
}
```

Visualisation of RNA-Seq reads. Take avg across replicates, smoothen it using rolling avg, and visualise as line graph.

```{r RNASeqvisualisation}

tmplength=length(WTRegionreads[,1])

#Avg of normalised counts, smoothing
WTregionavg=rowMeans(WTRegionreads[,3:(2+WTreps)])
WTregionavg=WTregionavg*10^6
KOregionavg=rowMeans(KORegionreads[,3:(2+KOreps)])
KOregionavg=KOregionavg*10^6

WTregionavg=rollmean(WTregionavg,k=400,fill="extend")
KOregionavg=rollmean(KOregionavg,k=400,fill="extend")

#Setting up visualisation matrix
WTregionvisual=as.data.frame(matrix(nrow=tmplength,ncol=3))
colnames(WTregionvisual)=c("Type","Loc","Normsmoothreads")
WTregionvisual$Type="WT"
WTregionvisual$Loc=WTRegionreads$Loc
WTregionvisual$Normsmoothreads=WTregionavg

KOregionvisual=as.data.frame(matrix(nrow=tmplength,ncol=3))
colnames(KOregionvisual)=c("Type","Loc","Normsmoothreads")
KOregionvisual$Type="KO"
KOregionvisual$Loc=KORegionreads$Loc
KOregionvisual$Normsmoothreads=KOregionavg

visualisationmat=rbind(WTregionvisual,KOregionvisual)
ggplot(visualisationmat,aes(x=Loc,y=Normsmoothreads,color=Type)) + geom_line(size=1.25)  + theme_bw() + theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())+scale_color_manual(values=c("black","grey70"))

#ggplot(visualisationmat,aes(x=Loc,y=Normsmoothreads,color=Type)) + geom_line(size=0.5)  + theme_bw() + theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())+scale_color_manual(values=c("red","darkblue"))
```

Normalise IP signal using WCE. Take care around infinite values generated by low WCE (set to maximum enrichment seen elsewhere), and obtain average enrichment (from non-infinite values). Do for both reduced and optimal salt samples.

```{r chipseqenrichment}
Normalised_reducedIP=Reduced_IP
Normalised_reducedIP$V3=Normalised_reducedIP$V3/Reduced_WCE$V3
rm(Reduced_IP)
rm(Reduced_WCE)

Normalised_optimalIP=Optimal_IP
Normalised_optimalIP$V3=Normalised_optimalIP$V3/Optimal_WCE$V3
rm(Optimal_IP)
rm(Optimal_WCE)

maxenrichment_reduced=max(Normalised_reducedIP$V3[is.finite(Normalised_reducedIP$V3)])
avgenrichment_reduced=mean(Normalised_reducedIP$V3[is.finite(Normalised_reducedIP$V3)])

maxenrichment_optimal=max(Normalised_optimalIP$V3[is.finite(Normalised_optimalIP$V3)])
avgenrichment_optimal=mean(Normalised_optimalIP$V3[is.finite(Normalised_optimalIP$V3)])

l2=length(Normalised_reducedIP$V1)
for (i in (1:l2)) {
  if (is.infinite(Normalised_reducedIP$V3[i])) {
    Normalised_reducedIP$V3[i]=maxenrichment_reduced
  }
  if (is.infinite(Normalised_optimalIP$V3[i])) {
    Normalised_optimalIP$V3[i]=maxenrichment_optimal
  }
}

```


Restrict the data to the relevant region. Smoothen the curve for visualisation using rolling means. Equalise the baselines of reduced and optimal samples to allow easy simultaneous visualisation. 

```{r Chipseqvisualisation}

#Restricting coordinates to genes and neighbouring regions
RestrictednormalisedIP_reduced=Normalised_reducedIP
RestrictednormalisedIP_optimal=Normalised_optimalIP

cs_startindex=which(RestrictednormalisedIP_reduced$V2==Regionstart)
cs_startindex=cs_startindex-buffer
cs_stopindex=which(RestrictednormalisedIP_reduced$V2==Regionstop)
cs_stopindex=cs_stopindex+buffer

RestrictednormalisedIP_reduced=RestrictednormalisedIP_reduced[c(cs_startindex:cs_stopindex),]
RestrictednormalisedIP_optimal=RestrictednormalisedIP_optimal[c(cs_startindex:cs_stopindex),]

#Smoothen curve, setup  matrix for visualisation -- reduced salt data
RestrictednormalisedIP_reduced$V3=rollmean(RestrictednormalisedIP_reduced$V3,k=400,fill="extend")
RestrictednormalisedIP_reduced$V1="Reduced"

#Smoothen curve, setup matrix for visualisation, equalise baselines -- optimal salt data
RestrictednormalisedIP_optimal$V3=rollmean(RestrictednormalisedIP_optimal$V3,k=400,fill="extend")
baselinediff=avgenrichment_reduced - avgenrichment_optimal
RestrictednormalisedIP_optimal$V3=RestrictednormalisedIP_optimal$V3+baselinediff
RestrictednormalisedIP_optimal$V1="Optimal"

#Combine optimal and reduced salt matrices in one to allow ggplot use
Restricted_normalisedIP_combined=rbind(RestrictednormalisedIP_reduced,RestrictednormalisedIP_optimal)

#Graphing code
#ggplot(Restricted_normalisedIP_combined,aes(x=V2,y=V3,fill=V1)) + geom_line(size=0.5,aes(color=V1)) + geom_ribbon(aes(ymin=min(V3),ymax=V3,alpha=0.2)) + scale_color_manual(values=c("grey70","grey30")) + scale_fill_manual(values=c("grey70","grey30")) + geom_hline(yintercept=avgenrichment_reduced,size=1.5,linetype="dotted") + theme_bw() + theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())

ggplot(Restricted_normalisedIP_combined,aes(x=V2,y=V3,fill=V1)) + geom_line(size=0.5,aes(color=V1)) + geom_ribbon(aes(ymin=min(V3),ymax=V3,alpha=0.01)) + scale_color_manual(values=c("darkblue","coral")) + scale_fill_manual(values=c("darkblue","coral")) + geom_hline(yintercept=avgenrichment_reduced,size=1.5,linetype="dotted") + theme_bw() + theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())
```


