---
title: "Mosaicsbinning_visualisation_combined"
output: html_document
---

## Instructions: Input a sorted BAM file, using Mosaics to generate binned read depth data. Then input the bins generated by Mosaics. Also read in name of main chromosome to be visualised. Also input 4 regions to be excluded from visualisation: hpyA and ura3 loci which have uneven coverage (at 112 and 1246 kb respectively), and two peaks present in HA tag-alone data (at 166 and 1710kb).

BAM files too large to upload to github. Please access these raw FastQ files via GEO accession GSE182514 and then use code in File S1 on github repository to generate BAME files for input https://github.com/amyschmid/HpyA_codes.

## This code generates visualization of ChIP-seq peaks for Figure 3A.


```{r libraries}
library(mosaics)
library(bazar)
```

```{r BAM_iput}
HA_BAM_WCE="HA_CM_WCE_stat"
HA_BAM_IP="HA_M_IP_stat"

HpyA_CMlog_BAM_WCE="hpyA_CM1_WCE_log"
HpyA_CMlog_BAM_IP="hpyA_CM2_IP_log"
  
HpyA_Mlog_BAM_WCE="hpyA_M3_WCE_log"
HpyA_Mlog_BAM_IP="hpyA_M3_IP_log"

HpyA_CMstat_BAM_WCE="hpyA_CM3_WCE_stat"
HpyA_CMstat_BAM_IP="hpyA_CM3_IP_stat"

HpyA_Mstat_BAM_WCE="hpyA_M4_WCE_stat"
HpyA_Mstat_BAM_IP="hpyA_M4_IP_stat"
  
fraglength=250
```


Constructing bins from BAM files using Mosaics
```{r construct_bins, results="hide", message=FALSE}
dir.create("Bins")

constructBins(infile=paste(HA_BAM_WCE,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)
constructBins(infile=paste(HA_BAM_IP,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)

constructBins(infile=paste(HpyA_CMlog_BAM_WCE,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)
constructBins(infile=paste(HpyA_CMlog_BAM_IP,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)

constructBins(infile=paste(HpyA_Mlog_BAM_WCE,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)
constructBins(infile=paste(HpyA_Mlog_BAM_IP,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)

constructBins(infile=paste(HpyA_CMstat_BAM_WCE,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)
constructBins(infile=paste(HpyA_CMstat_BAM_IP,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)

constructBins(infile=paste(HpyA_Mstat_BAM_WCE,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)
constructBins(infile=paste(HpyA_Mstat_BAM_IP,"_sorted.bam",sep=""),fileFormat="bam",outfileLoc="./Bins",byChr=FALSE,fragLen=fraglength,binSize=fraglength,capping=0,PET=FALSE)
```

Read in constructed Bins
```{r read_bins}
setwd('./Bins')

HA_IP=read.table(paste(HA_BAM_IP,"_sorted.bam_fragL250_bin250.txt",sep=""))
HA_WCE=read.table(paste(HA_BAM_WCE,"_sorted.bam_fragL250_bin250.txt",sep=""))

Logopt_IP=read.table(paste(HpyA_CMlog_BAM_IP,"_sorted.bam_fragL250_bin250.txt",sep=""))
Logopt_WCE=read.table(paste(HpyA_CMlog_BAM_WCE,"_sorted.bam_fragL250_bin250.txt",sep=""))

Loglow_IP=read.table(paste(HpyA_Mlog_BAM_IP,"_sorted.bam_fragL250_bin250.txt",sep=""))
Loglow_WCE=read.table(paste(HpyA_Mlog_BAM_WCE,"_sorted.bam_fragL250_bin250.txt",sep=""))

Statopt_IP=read.table(paste(HpyA_CMstat_BAM_IP,"_sorted.bam_fragL250_bin250.txt",sep=""))
Statopt_WCE=read.table(paste(HpyA_CMstat_BAM_WCE,"_sorted.bam_fragL250_bin250.txt",sep=""))

Statlow_IP=read.table(paste(HpyA_Mstat_BAM_IP,"_sorted.bam_fragL250_bin250.txt",sep=""))
Statlow_WCE=read.table(paste(HpyA_Mstat_BAM_WCE,"_sorted.bam_fragL250_bin250.txt",sep=""))

chrname="NC_002607.1"

HApeak1=c(166000,167000)
HApeak2=c(1710750,1711250)
Unevencoverage1=c(110250,112750)
Unevencoverage2=c(1246000,1247000)

setwd('../')
```

Restrict data to main chromosome, get length and location of each gene.
```{r preliminaries}

HA_IP=HA_IP[HA_IP$V1==paste(chrname),]
HA_WCE=HA_WCE[HA_WCE$V1==paste(chrname),]

Logopt_IP=Logopt_IP[Logopt_IP$V1==paste(chrname),]
Logopt_WCE=Logopt_WCE[Logopt_WCE$V1==paste(chrname),]

Loglow_IP=Loglow_IP[Loglow_IP$V1==paste(chrname),]
Loglow_WCE=Loglow_WCE[Loglow_WCE$V1==paste(chrname),]

Statlow_IP=Statlow_IP[Statlow_IP$V1==paste(chrname),]
Statlow_WCE=Statlow_WCE[Statlow_WCE$V1==paste(chrname),]

```

Normalise IP wignal using WCE
```{r chipseqenrichment}

HA_NormalisedIP=HA_IP
HA_NormalisedIP$V3=HA_NormalisedIP$V3/HA_WCE$V3

Logopt_NormalisedIP=Logopt_IP
Logopt_NormalisedIP$V3=Logopt_NormalisedIP$V3/Logopt_WCE$V3

Loglow_NormalisedIP=Loglow_IP
Loglow_NormalisedIP$V3=Loglow_NormalisedIP$V3/Loglow_WCE$V3

Statopt_NormalisedIP=Statopt_IP
Statopt_NormalisedIP$V3=Statopt_NormalisedIP$V3/Statopt_WCE$V3

Statlow_NormalisedIP=Statlow_IP
Statlow_NormalisedIP$V3=Statlow_NormalisedIP$V3/Statlow_WCE$V3
```

Removing regions of uneven coverage - hpyA and ura3 loci - that skew results. Also removing peaks found in HA tag-alone negative control to better highlight real peaks
```{r cleanup}
#l=length*HA_NormalisedIP$V1

#Indices of HA peaks/uneven coverage regions
HA1index1=which(HA_NormalisedIP$V2==HApeak1[1])
HA1index2=which(HA_NormalisedIP$V2==HApeak1[2])
HA2index1=which(HA_NormalisedIP$V2==HApeak2[1])
HA2index2=which(HA_NormalisedIP$V2==HApeak2[2])
UC1index1=which(HA_NormalisedIP$V2==Unevencoverage1[1])
UC1index2=which(HA_NormalisedIP$V2==Unevencoverage1[2])
UC2index1=which(HA_NormalisedIP$V2==Unevencoverage2[1])
UC2index2=which(HA_NormalisedIP$V2==Unevencoverage2[2])

#Acg enrichment across the chromsome - replace funky values with this one.
HA_avgenrichment=mean(HA_NormalisedIP$V3[is.finite(HA_NormalisedIP$V3)])
Logopt_avgenrichment=mean(Logopt_NormalisedIP$V3[is.finite(Logopt_NormalisedIP$V3)])
Loglow_avgenrichment=mean(Loglow_NormalisedIP$V3[is.finite(Loglow_NormalisedIP$V3)])
Statopt_avgenrichment=mean(Statopt_NormalisedIP$V3[is.finite(Statopt_NormalisedIP$V3)])
Statlow_avgenrichment=mean(Statlow_NormalisedIP$V3[is.finite(Statlow_NormalisedIP$V3)])

#Replcament of high values caused by HA binding or local depletion with avg genome-wide value.
HA_NormalisedIP2=HA_NormalisedIP
HA_NormalisedIP2$V3[UC1index1:UC1index2]=HA_avgenrichment
HA_NormalisedIP2$V3[UC2index1:UC2index2]=HA_avgenrichment

HA_NormalisedIP3=HA_NormalisedIP2
HA_NormalisedIP3$V3[HA1index1:HA1index2]=HA_avgenrichment
HA_NormalisedIP3$V3[HA2index1:HA2index2]=HA_avgenrichment

Logopt_NormalisedIP$V3[HA1index1:HA1index2]=Logopt_avgenrichment
Logopt_NormalisedIP$V3[HA2index1:HA2index2]=Logopt_avgenrichment
Logopt_NormalisedIP$V3[UC1index1:UC1index2]=Logopt_avgenrichment
Logopt_NormalisedIP$V3[UC2index1:UC2index2]=Logopt_avgenrichment

Loglow_NormalisedIP$V3[HA1index1:HA1index2]=Loglow_avgenrichment
Loglow_NormalisedIP$V3[HA2index1:HA2index2]=Loglow_avgenrichment
Loglow_NormalisedIP$V3[UC1index1:UC1index2]=Loglow_avgenrichment
Loglow_NormalisedIP$V3[UC2index1:UC2index2]=Loglow_avgenrichment

Statopt_NormalisedIP$V3[HA1index1:HA1index2]=Statopt_avgenrichment
Statopt_NormalisedIP$V3[HA2index1:HA2index2]=Statopt_avgenrichment
Statopt_NormalisedIP$V3[UC1index1:UC1index2]=Statopt_avgenrichment
Statopt_NormalisedIP$V3[UC2index1:UC2index2]=Statopt_avgenrichment

Statlow_NormalisedIP$V3[HA1index1:HA1index2]=Statlow_avgenrichment
Statlow_NormalisedIP$V3[HA2index1:HA2index2]=Statlow_avgenrichment
Statlow_NormalisedIP$V3[UC1index1:UC1index2]=Statlow_avgenrichment
Statlow_NormalisedIP$V3[UC2index1:UC2index2]=Statlow_avgenrichment

```

Plot ChIP-Seq depth (IP enrichment over WCE)
```{r plotting,fig.height = 3.2, fig.width = 7}

plot(x=HA_NormalisedIP$V2,HA_NormalisedIP$V3,type="h",col="black",ylim=c(0,3))

plot(x=HA_NormalisedIP2$V2,HA_NormalisedIP2$V3,type="h",col="black",ylim=c(0,3))

plot(x=HA_NormalisedIP3$V2,HA_NormalisedIP3$V3,type="h",col="black",ylim=c(0,3))

plot(x=Logopt_NormalisedIP$V2,Logopt_NormalisedIP$V3,type="h",col="red",ylim=c(0,3))

plot(x=Loglow_NormalisedIP$V2,Loglow_NormalisedIP$V3,type="h",col="darkblue",ylim=c(0,3))

plot(x=Statopt_NormalisedIP$V2,Statopt_NormalisedIP$V3,type="h",col="lightcoral",ylim=c(0,3))

plot(x=Statlow_NormalisedIP$V2,Statlow_NormalisedIP$V3,type="h",col="cornflowerblue",ylim=c(0,3))
```
